### PR TITLE
enhancement(codecs)!: Add support for enhanced metric tags to `native_json`

### PR DESCRIPTION
**Pull request by bruceg**
This is stacked on #15286 and will be updated when that is merged.

Ref #14742

### CODE REVIEW

File: lib/vector-core/src/event/metric/tags.rs [hunk 1]
```diff
@@ -377,18 +377,38 @@ impl ByteSizeOf for TagValueSet {
 
 impl<'de> Deserialize<'de> for TagValueSet {
     fn deserialize<D: Deserializer<'de>>(de: D) -> Result<Self, D::Error> {
-        // Deserialize from a single string only
-        let s = String::deserialize(de)?;
-        Ok(Self::from([s]))
+        #[derive(Deserialize)]
+        #[serde(untagged)]
+        enum Variants {
```
Comment (neuronull) [inline lib/vector-core/src/event/metric/tags.rs:382]: Just for curiosity and I'm guessing the answer is going to be "no" - are there any performance advantages to declaring this enum outside the function body ?

Comment (bruceg) [reply lib/vector-core/src/event/metric/tags.rs:382]: Most likely no. The only possible differences I can imagine is if the change in the localization affected either inlining or code layout in a predictable way.

**Review by spencergilbert — APPROVED** [review]

File: lib/vector-core/src/event/metric/tags.rs [hunk 2]
```diff
@@ -5,12 +5,13 @@ use std::hash::{Hash, Hasher};
...
 
 use indexmap::IndexSet;
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde::{ser::SerializeSeq, Deserialize, Deserializer, Serialize, Serializer};
 use vector_common::byte_size_of::ByteSizeOf;
 use vector_config::{configurable_component, Configurable};
 
 /// A single tag value, either a bare tag or a value.
 #[derive(Clone, Configurable, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[serde(untagged)]
```
Comment (tobz) [inline lib/vector-core/src/event/metric/tags.rs:14]: Hmm, is my intuition correct here that, with the change to be untagged... a null value for a field of type `TagValue` will get deserialized as `TagValue::Bare`?

**Review by tobz — APPROVED** [review]
This looks straightforward enough to me. 👍🏻 

I just left one non-blocking question.

File: lib/vector-core/src/event/metric/tags.rs (hunk 2 above)
Comment (bruceg) [reply lib/vector-core/src/event/metric/tags.rs:14]: That is the intention, yes.

**Pull request merged**